content/blog/2023/contain-root.md

```diff
@@ -0,0 +1,517 @@
+- There are no outside counters or quotes
+  for us to worry about.
+
+With `container-type: size` on the root,
+we could replace most or all
+`@media` size queries with `@container`
+across our sites.
+I think we should!
+
+## Is it safe to contain the root element?
+
+👍🏼 It should be. \
+👎🏼 But it isn't. \
+👍🏼 Unless you're careful.
+
+(Hopefully browsers will get this fixed.)
+
+Currently,
+adding a `container-type` of `inline-size`
+to the root (`html`) element
+will make scrolling impossible:
+
+{{ embed.codepen(
+  id='dyQWdag',
+  title='HTML containment & overflow (1d)',
+  user='miriamsuzanne'
+) }}
+
+And using `inline-size` instead
+will fix the scrollbar issue,
+but now fixed elements
+scroll off the page:
+
+{{ embed.codepen(
+  id='QWJbLLe',
+  title='HTML containment & overflow (2d)',
+  user='miriamsuzanne'
+) }}
+
+Neither behavior
+will work as a default for most sites.
+Thanks to [Temani Afif](https://front-end.social/@css/110513977760791600)
+for initially pointing out the issue
+on Mastodon.
+
+Various people in that thread
+pointed to 'root/body propagation'
+as the cause.
+And they're right --
+but as far as I can tell,
+they shouldn't be.
+I was involved with several conversations
+in the CSS Working Group
+around containment and propagation,
+and this is not what we decided.
+It's also not defined
+to work like this
+in the specification.
+
+As far as I can tell,
+this is a browser bug --
+implemented in all browsers --
+that needs to be fixed.
+
+## What is root/body propagation, and how does it work?
+
+For us as web authors,
+the root of a document
+is the `html` element.
+But for browsers,
+there's more context to worry about
+-- more root than the root --
+such as
+the _viewport_,
+the _document canvas_,
+the _initial containing block_,
+and the _initial layout block_.
+
+I am not an expert on the complexities
+of these browser concepts --
+but roughly, together,
+they describe the context
+immediately surrounding our web pages
+in the browser:
+the 'viewport' through which we're looking,
+the 'canvas' our site is painted onto,
+and the 'blocks' that defines
+our initial positioning and layout context.
+For simplicity,
+I'm going to refer to all of this
+collectively as The Viewport™️.
+
+We don't have direct access
+to The Viewport™️.
+There's no CSS syntax we can use
+to select and style any aspect of The Viewport™️.
+And yet, we style them all the time
+using the somewhat quirky and esoteric magic
+of 'propagation' --
+where styles on one element
+(in this case `html` or `body`)
+bubble outward ('propagate')
+and apply to a parent element instead.
+
+For example,
+when we set a `background` on `html`,
+that background
+[propagates to the document canvas](https://drafts.csswg.org/css-backgrounds/#special-backgrounds)
+instead.
+If we don't set a background on `html`,
+but we do set a background on `body`,
+then the `body` background
+propagates instead --
+skipping over the root,
+and straight up to the canvas!
```

Comment:
Wow, TIL! Surprised of the amount of little edge cases like this browsers have to support